smart-contracts/test/helpers/uniswapV3.js

```diff
@@ -320,12 +322,96 @@ const deployUniswapV3Oracle = async function () {
   return oracle
 }
 
+
+async function getUniswapRoute (
+  tokenIn, 
+  tokenOut, 
+  amoutOut = ethers.utils.parseUnits('10', tokenOut.decimals),
+  recipient,
+  chainId = 1
+) {
+  // default to first signer
+  if(!recipient) {
+    const [signer] = await ethers.getSigners()
+    recipient = signer.address
+  }
+
+  console.log(`Swap ${tokenIn.symbol} > ${tokenOut.symbol}`)
+  console.log(`Out: ${amoutOut} to ${recipient}`)
+
+  // init router
+  const router = new AlphaRouter({ 
+    chainId, 
+    provider: ethers.provider,
+  })
+
+  // parse router args 
+  const outputAmount = CurrencyAmount.fromRawAmount(tokenOut, JSBI.BigInt(amoutOut))
+  const args = {
+    outputAmount,
+    quoteCurrency: tokenIn,
+    swapType: TradeType.EXACT_OUTPUT,
+    swapConfig: {
+      type: SwapType.UNIVERSAL_ROUTER,
+      recipient,
+      slippageTolerance: new Percent(5, 100),
+      deadline: Date.now() + 1800
+    }
+  }
+
+  // call router
+  const route = await router.route(
+    ...Object.values(args)
+  )
+
+  // log some prices
+  console.log(`Quote Exact In: ${route.quote.toFixed(2)}`);
+  console.log(`Gas Adjusted Quote In: ${route.quoteGasAdjusted.toFixed(2)}`);
+  console.log(`Gas Used USD: ${route.estimatedGasUsedUSD.toFixed(6)}`);
+
+  const { methodParameters, quote } = route
+  const { 
+      calldata: swapCalldata, 
+      value, 
+      to: swapRouter
+  } = methodParameters
+
+  // parse quote as BigNumber
+  const amountInMax = ethers.BigNumber.from(
+    JSBI.multiply(
+      quote.numerator, 
+      quote.denominator,
+      quote.decimalScale
+    ).toString()
+  )
```

Comment:
it seems this calculation is wrong, which lead the erc20 > native swaps to fail. Couldn't figure out a simple way to convert Uniswap `CurrencyAmount` into ethers BigNumber... annoying.

Here is a simple script that seemed to work (but uniswap values are more complex and fails)

```
const { ethers } = require('hardhat')
const { getUniswapTokens } = require('../../test/helpers')
const { CurrencyAmount } = require('@uniswap/sdk-core')
const JSBI  = require('jsbi')

async function main() {
  
  const tokens = getUniswapTokens()

  const token = tokens.native
  const bn = ethers.utils.parseUnits('119727.47955', token.decimals)
  const outputAmount = CurrencyAmount.fromRawAmount(
    token, 
    JSBI.BigInt(bn)
  )

  console.log(outputAmount)
  console.log(outputAmount.toExact())
  console.log(outputAmount.toFixed(2))
  console.log(outputAmount.toSignificant())
  
  const str = JSBI.multiply(
    outputAmount.numerator, 
    outputAmount.denominator,
    outputAmount.decimalScale
  ).toString()

  console.log(bn.toString())
  console.log(str)
  
  const re = ethers.BigNumber.from(str)
  console.log(re.toString())
}

main()
  .then(() => process.exit(0))
  .catch((error) => {
    console.error(error)
    process.exit(1)
  })
```
```

Reply:
ok finally resolved this - and discovered the existence of `FixedNumber`...